### PR TITLE
Search endpoint

### PR DESCRIPTION
**Pull request by PratikMahajan**
Search endpoint with categories, tags and query

### CODE REVIEW

File: handlers/apps.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+
+	"github.com/knative-scout/app-api/models"
+	"go.mongodb.org/mongo-driver/bson"
+	"net/http"
+	"strings"
+)
+
+// HealthHandler is used to determine if the server is running
```
Comment (catvec) [inline handlers/apps.go]: Update doc comment.

File: handlers/apps.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	BaseHandler
+}
+
+// ServeHTTP implements http.Handler
+func (h AppsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	//gets all the optional parametres passed in the URL
+	vars := r.URL.Query()
+
+	//h.Logger.Debugf("%#v", vars)
```
Comment (catvec) [inline handlers/apps.go]: delete line

File: handlers/apps.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	query:=vars.Get("query")
```
Comment (catvec) [inline handlers/apps.go]: add space after =

File: handlers/apps.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	tags:= vars.Get("tags")
+	categories:=vars.Get("categories")
```
Comment (catvec) [inline handlers/apps.go]: add space after =

File: handlers/apps.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+	resp := getDataFromDB(query, tags, categories, h)
+
+	h.RespondJSON(w, http.StatusOK, resp)
+}
+
+
```
Comment (catvec) [inline handlers/apps.go]: Add doc comment

File: handlers/apps.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	if len(categories)>0{
+		categories := strings.Split(categories, ",")
+		searchBson = append(searchBson, bson.E{"categories", bson.D{{"$in", categories}}})
+	}
+
+
+	ret := []models.App{}
+	result, err := h.MDbApps.Find(h.Ctx, searchBson)
+
+	if err != nil {
```
Comment (catvec) [inline handlers/apps.go]: If an error occurs, the handler as to exit so the user knows an error occurred.

panic this.

File: handlers/apps.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+		 h.Logger.Fatalf("failed to retrieve data from db %s", err)
+	}
+
+	for result.Next(h.Ctx) {
+		a := models.App{}
+		if err = result.Decode(&a); err != nil {
+			h.Logger.Fatalf("readTasks: couldn't make to-do item ready for display: %v", err)
```
Comment (catvec) [inline handlers/apps.go]: ditto

File: main.go [hunk 8]
```diff
@@ -156,11 +156,21 @@ func main() {
...
+
+	// Endpoint :"/health"
 	router.Handle("/health", handlers.HealthHandler{
 		baseHandler.GetChild("health"),
 	}).Methods("GET")
 
+	// Endpoint :"/search"
+	router.Handle("/apps", handlers.AppsHandler{
+		baseHandler.GetChild("apps"),
+	}).Methods("GET")//.Queries("query", "{[0-9]*?}","tags","{[0-9]*?}")
```
Comment (catvec) [inline main.go]: remove commented out

**Review by catvec — CHANGES_REQUESTED** [review]

File: handlers/apps.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+	//gets all the optional parameters passed in the URL
+	vars := r.URL.Query()
+	query := vars.Get("query")
+	tags := vars.Get("tags")
+	categories := vars.Get("categories")
+
+	resp := getDataFromDB(query, tags, categories, h)
+
+	h.RespondJSON(w, http.StatusOK, resp)
```
Comment (catvec) [inline handlers/apps.go]: Can you make it return data according to the spec? https://github.com/knative-scout/app-api/blob/master/DESIGN.md#search-apps

Should return an object with an `apps` key which should then holds the apps you've found

File: handlers/apps.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/knative-scout/app-api/models"
+	"go.mongodb.org/mongo-driver/bson"
+	"net/http"
+	"strings"
+)
+
+// AppsHandler is used search apps and return result
+// in case of an empty query, it returns all the apps in the database
+
+type AppsHandler struct {
```
Comment (catvec) [inline handlers/apps.go]: Can you name this something less general. B/c we will have a few more app endpoints.

Maybe `AppSearchHandler`?

Then rename the `handlers/apps.go` file to something less general like `handlers/search,go`

**Review by catvec — CHANGES_REQUESTED** [review]

File: handlers/apps.go (hunk 10 above)
Comment (PratikMahajan) [reply handlers/apps.go]: okay

File: handlers/apps.go (hunk 9 above)
Comment (PratikMahajan) [reply handlers/apps.go]: okay

**Review by catvec — APPROVED** [review]

**Pull request merged**